src/sources/mod.rs

```diff
@@ -157,25 +164,28 @@ pub trait EventSource {
     /// [`Poll::unregister`](crate::Poll#method.unregister) method.
     fn unregister(&mut self, poll: &mut Poll) -> crate::Result<()>;
 
-    /// Notification that a polling session is going to start
+    /// Whether this source needs to be sent the [`EventSource::before_will_sleep`]
+    /// and [`EventSource::before_handle_events`] notifications. These are opt-in because
+    /// they require more expensive checks, and almost all sources will not need these notifications
+    const NEEDS_EXTRA_LIFECYCLE_EVENTS: bool = false;
+    /// Notification that a single `poll` is about to begin.
+    /// If this returns Ok(Some), polling will shortcircuit, and
+    /// your event handler will be called with the returned Token and Readiness
     ///
-    /// You can generate events from this method as you would from `process_events`.
-    fn pre_run<F>(&mut self, _callback: F) -> crate::Result<()>
-    where
-        F: FnMut(Self::Event, &mut Self::Metadata) -> Self::Ret,
-    {
-        Ok(())
-    }
-
-    /// Notification that the current polling session ended
+    /// Use this to perform operations which must be done before polling,
+    /// but which may conflict with other event handlers. For example,
+    /// if polling requires a lock to be taken.
+    fn before_will_sleep(&mut self) -> crate::Result<Option<(Readiness, Token)>> {
```

Comment:
What about naming this `before_sleep` instead? This matches better the naming pattern of the other method.